enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,584 @@
+IngressController resource is generated as exepected.
+
+- Add periodic nightly tests which install single-node in the cloud, add a few
+worker nodes to it, then run conformance tests to make sure we don't run into
+any problems not described in this enhancement.
+
+- Add periodic nightly tests which install a single-node "none"-platform cluster,
+add worker nodes to it, and check that ingress traffic still works as expected
+and recovers even after the `router-default` pod gets deleted and rescheduled.
+Make sure this is still true even after upgrades.
+
+- Add tests on both cloud / "none"-platform that check that a single-node cluster
+with additional workers recovers after the single control-plane node reboots by
+running conformance tests post-reboot.
+
+- Add tests on both cloud / "none"-platform that check that a single-node cluster
+with additional workers recovers after an upgrade by running conformance tests
+post-upgrade.
+
+TODO: Describe day-1-workers tests?
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+TBD
+
+#### Dev Preview -> Tech Preview
+
+TBD
+
+#### Tech Preview -> GA
+
+TBD
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+The new `DefaultPlacement` API field will have a default value of `Workers`.
+The behavior of the Cluster Ingress Operator when the `DefaultPlacement` field
+has the value `Workers` should be identical to its current behavior (before
+this enhancement). This means that clusters that go through an upgrade from a
+version before this enhancement to a later version which includes this
+enhancement will maintain their current behavior, this enhancement should not
+change anything for them.
+
+If the value of `DefaultPlacement` is empty (TODO: Could this possibly happen
+if the the Cluster Ingress Operator reads a resource created according to the
+old Ingress CRD that didn't have this value? Not sure how defaults work in this
+scenario, in any case, it's not crucial) the Cluster Ingress Operator should
+make sure to treat it as if it were to have the value `Workers`.
+
+### Version Skew Strategy
+
+Does not apply, to the best of my understanding.
+
+### Operational Aspects of API Extensions
+
+- Since the `DefaultPlacement` field is part of the status of the
+Ingress config CR, it is clear that users should not modify this field.
+
+- This API change only affects the defaulting behavior of the `IngressController`
+CR, it does not add any new capabilities to OCP, or give any more flexibility
+than there already was.
+
+- For administrators and support engineers, the `IngressController` is still the
+source of truth and where you need to look if you seek to understand the router
+placement in practice. However, now the `replicas` and `nodePlacement` on
+`IngressController` may be empty, so administrators and support engineers 
+can refer to the corresponding Deployment that got created if they wish to
+know the practical values.
```

Comment:
Why wouldn't we have the IngressController fill in the settings it is using a default, so the user doesn't have to guess?

Reply:
Do you mean the ingress operator should actively reconcile all IngressController CRs all the time to populate these values when unset with the defaults that would otherwise be set in the corresponding Deployment?

I guess it could do that, I'm not the one to say how typical it is for an operator to do something like that. I guess this is something we can can discuss in the ingress PR? I don't have an opinion here

Reply:
Yes, that's what I meant.

@JoelSpeed @Miciah is this worth resolving here?

Reply:
(Added 9ef9aac54541c72b764019dd756b7c7f725f6980 to mention this idea)

Reply:
@omertuc and I discussed this on a call before, I thought there was a note already. Upshot of our conversation is that there are two ways you can default, either via webhook or via an operator at runtime. For this particular field, if we default via the operator at runtime, that allows us flexibility to change the defaults in the future (note the comment on the API about this). If we default via a webhook, whatever gets set on the resource is now there and set, we cannot change that default later because there will be existing users with the default encoded in their resource even though they had no opinion on it.

For this particular example, we agreed a runtime defaulting in the controller was likely the best approach. Adding a webhook defaulting option to the alternatives is a reasonable ask though IMO (where we can document why we aren't doing that)

I'm going to add a section talking about defaults to the API conventions doc I'm working on.

Reply:
@JoelSpeed my understanding is that what's being suggested in this thread is neither what you would call "webhook defaulting" nor "operator at runtime defaulting" - it's more like the operator continuously doing the defaulting work the webhook would otherwise do, by making sure those values are not left unset in the IngressControllers, but are instead set to their default values.

Reply:
That's right, one difference being that it overcomes the problem of the webhook not being present during installation time

Reply:
Yes, Omer's description of what I was asking about is correct.

If we want the default to be able to change, then leaving the values unset is the right way to tell the difference between values the user gives and the defaults. We could reflect the actual values being used in the status field to aid in debugging, but if those fields aren't already there then that's probably a different enhancement.

Omer, could you add the conclusion from this thread to the body of the doc and then resolve it, please?

Reply:
My conclusion remains what I've added in 9ef9aac54541c72b764019dd756b7c7f725f6980, do you think I missed any important detail?

Reply:
I think what you've written covers it adequately. Thanks!